src/Compilers/Core/AnalyzerDriver/AnalyzerExecutor.cs

```diff
@@ -27,6 +27,8 @@ internal class AnalyzerExecutor
         private readonly Func<DiagnosticAnalyzer, bool> _isCompilerAnalyzer;
         private readonly ImmutableDictionary<DiagnosticAnalyzer, object> _singleThreadedAnalyzerToGateMap;
 
+        private ImmutableDictionary<DiagnosticAnalyzer, TimeSpan> _analyzerExecutionTimeMapOpt;
```

Comment:
ConcurrentDictionary might be better here. I'm worried about the amount of churn in the ImmutableDictionary (which is mutating all the time when analyzers are running).